src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+        })
+    }
+
+    fn load_contract(&self, contract_name: &str) -> Result<Option<Contract>> {
+        let contract: Option<String> =
+            self.query_row(
+                "SELECT contract_data FROM contracts WHERE contract_name = ?",
+                &[contract_name],
+                |row| {
+                    row.get(0)
+                })?;
+        match contract {
+            None => Ok(None),
+            Some(ref contract) => Ok(Some(Contract::deserialize(contract)?))
+        }
+    }
+
+    fn execute<P>(&mut self, sql: &str, params: P) -> Result<usize>
+    where
+        P: IntoIterator,
+        P::Item: ToSql {
+        match self.conn.take() {
+            Some(conn) => {
+                let result = conn.execute(sql, params)
+                    .map_err(|x| {
+                        eprintln!("SQL Execution Error: {:?}", x);
+                        Error::new(ErrType::SqliteError(IncomparableError{ err: x }))
+                    })?;
+                // if execution error'ed, return _without_ replacing connection!
+                //    this closes the connection.
+                self.conn.replace(conn);
+                Ok(result)
+            },
+            None => {
+                Err(Error::new(ErrType::SqlConnectionClosed))
+            }
+        }
+    }
+
+    fn query_row<T, P, F>(&self, sql: &str, params: P, f: F) -> Result<Option<T>>
+    where
+        P: IntoIterator,
+        P::Item: ToSql,
+        F: FnOnce(&Row) -> T {
+        if let Some(ref conn) = self.conn {
+            conn.query_row(sql, params, f)
+                .optional()
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+}
+
+impl ContractDatabase for SqliteContractDatabase {
+    fn create_map(&mut self, contract_name: &str, map_name: &str, key_type: TupleTypeSignature, value_type: TupleTypeSignature) -> Result<()> {
+        let key_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(key_type));
+        let value_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(value_type));
+
+        self.execute("INSERT INTO maps_table (contract_name, map_name, key_type, value_type) VALUES (?, ?, ?, ?)",
+                     &[contract_name, map_name, &key_type.serialize()?, &value_type.serialize()?])?;
+        Ok(())
+    }
+
+    fn fetch_entry(&self, contract_name: &str, map_name: &str, key: &Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), (*key).clone())))
+        }
+
+        let params: [&ToSql; 2] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?];
+
+        let sql_result: Option<Option<String>> = 
+            self.query_row(
```

Comment:
Did you want to add `LIMIT 1` to this query so it doesn't load excessive data?

Reply:
Sure, though I think query_row won't read more than 1 row from the database anyways.